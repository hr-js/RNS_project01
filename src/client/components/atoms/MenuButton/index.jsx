// @flow
import * as React from "react";
import styles from "./style.css";

type PropsType = {
  children: string,
  // flowのtype定義なのにreactのルールが適用されているので無効化
  // eslint-disable-next-line react/require-default-props
  selected?: string,
  onButtonClick: () => void
};

function MenuButton(props: PropsType): React.Node {
  const { selected } = props;
  const className: string =
    selected === "selected"
      ? `${styles.button} ${styles.selected}`
      : styles.button;
  return (
    <button className={className} onClick={props.onButtonClick}>
      {props.children}
    </button>
  );
}
export default MenuButton;
