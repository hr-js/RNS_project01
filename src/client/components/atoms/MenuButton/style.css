@charset "utf-8";
.button {
  width: 100%;
  height: 85px;
  margin: 0;
  padding: 0;
  font-size: 18px;
  color: gray;
  background: transparent;
  border: none;
  border-bottom: 1px dashed silver;
  box-sizing: border-box;
  transition: font-size 200ms;
  outline: none;
}
.button.selected {
  background-color: silver;
  border: none;
  color: white;
  font-size: 24px;
  font-weight: bold;
}
