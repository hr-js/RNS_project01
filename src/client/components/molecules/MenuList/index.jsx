// @flow
import * as React from "react";
import MenuButton from "@components/atoms/MenuButton";
import styles from "./style.css";

type PropType = {
  current: string,
  onButtonClick: (current: string) => void
};

const items: Array<string> = ["入退室処理", "ユーザ登録", "利用者一覧"];

function MenuList(props: PropType): React.Node {
  const { current, onButtonClick } = props;

  const listItems: Array<React.Node> = items.map((item: string): React.Node => {
    const menuButton: React.Node =
      item === current ? (
        <MenuButton
          selected="selected"
          onButtonClick={() => {
            onButtonClick(item);
          }}
        >
          {item}
        </MenuButton>
      ) : (
        <MenuButton
          onButtonClick={() => {
            onButtonClick(item);
          }}
        >
          {item}
        </MenuButton>
      );
    return <li key={item}>{menuButton}</li>;
  });

  return <ul className={styles.menuList}>{listItems}</ul>;
}
export default MenuList;
