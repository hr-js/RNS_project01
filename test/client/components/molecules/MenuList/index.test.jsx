import React from "react";
import renderer from "react-test-renderer";
import { shallow, mount } from "enzyme";
import MenuList from "@components/molecules/MenuList";
import MenuButton from "@components/atoms/MenuButton";

describe("MenuList.jsxのテスト", () => {
  const func = () => {};

  describe("スナップショット", () => {
    it("renders corecctly", () => {
      const tree = renderer
        .create(<MenuList current="入退室処理" onButtonClick={func} />)
        .toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe("コンポーネントのテスト", () => {
    it("MenuButtonが3つある", () => {
      const list = shallow(
        <MenuList current="入退室処理" onButtonClick={func} />
      );
      expect(list.find(MenuButton)).toHaveLength(3);
      expect(list.find(MenuButton).get(0).props.children).toBe("入退室処理");
      expect(list.find(MenuButton).get(1).props.children).toBe("ユーザ登録");
      expect(list.find(MenuButton).get(2).props.children).toBe("利用者一覧");
    });

    describe("currentに「入退室処理」を指定する", () => {
      it("「入退室処理」ボタンにselected属性がある", () => {
        const list = shallow(
          <MenuList current="入退室処理" onButtonClick={func} />
        );
        expect(list.find(MenuButton).get(0).props.selected).toBe("selected");
        expect(list.find(MenuButton).get(1).props.selected).toBeUndefined();
        expect(list.find(MenuButton).get(2).props.selected).toBeUndefined();
      });
    });

    describe("currentに「ユーザ登録」を指定する", () => {
      it("「ユーザ登録」ボタンにselected属性がある", () => {});
      const list = shallow(
        <MenuList current="ユーザ登録" onButtonClick={func} />
      );
      expect(list.find(MenuButton).get(0).props.selected).toBeUndefined();
      expect(list.find(MenuButton).get(1).props.selected).toBe("selected");
      expect(list.find(MenuButton).get(2).props.selected).toBeUndefined();
    });

    describe("currentに「利用者一覧」を指定する", () => {
      it("「利用者一覧」ボタンにselected属性がある", () => {
        const list = shallow(
          <MenuList current="利用者一覧" onButtonClick={func} />
        );
        expect(list.find(MenuButton).get(0).props.selected).toBeUndefined();
        expect(list.find(MenuButton).get(1).props.selected).toBeUndefined();
        expect(list.find(MenuButton).get(2).props.selected).toBe("selected");
      });
    });

    describe("MenuButtonをクリック", () => {
      const onButtonClick = jest.fn();
      beforeEach(() => {
        onButtonClick.mockReset();
      });

      describe("「入退室処理」をクリックする", () => {
        it("引数に'入退室処理'が渡される", () => {
          const list = mount(
            <MenuList current="" onButtonClick={onButtonClick} />
          );
          list
            .find(MenuButton)
            .at(0)
            .simulate("click");
          expect(onButtonClick.mock.calls[0][0]).toBe("入退室処理");
        });
      });

      describe("「ユーザ登録」をクリックする", () => {
        it("引数に'ユーザ登録'が渡される", () => {
          const list = mount(
            <MenuList current="" onButtonClick={onButtonClick} />
          );
          list
            .find(MenuButton)
            .at(1)
            .simulate("click");
          expect(onButtonClick.mock.calls[0][0]).toBe("ユーザ登録");
        });
      });

      describe("「利用者一覧」をクリックする", () => {
        it("引数に'利用者一覧'が渡される", () => {
          const list = mount(
            <MenuList current="" onButtonClick={onButtonClick} />
          );
          list
            .find(MenuButton)
            .at(2)
            .simulate("click");
          expect(onButtonClick.mock.calls[0][0]).toBe("利用者一覧");
        });
      });
    });
  });
});
