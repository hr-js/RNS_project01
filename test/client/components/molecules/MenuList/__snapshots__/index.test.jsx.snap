// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MenuList.jsxのテスト スナップショット renders corecctly 1`] = `
<ul
  className="menuList"
>
  <li>
    <button
      className="button selected"
      onClick={[Function]}
    >
      入退室処理
    </button>
  </li>
  <li>
    <button
      className="button"
      onClick={[Function]}
    >
      ユーザ登録
    </button>
  </li>
  <li>
    <button
      className="button"
      onClick={[Function]}
    >
      利用者一覧
    </button>
  </li>
</ul>
`;
